flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/remote/RemoteExecutor.java

```diff
@@ -0,0 +1,423 @@
+        }
+        return ((MaterializedResult) result).snapshot(pageSize);
+    }
+
+    public List<RowData> retrieveResultPage(String resultId, int page) {
+        final DynamicResult result = resultStore.getResult(resultId);
+        if (result == null) {
+            throw new SqlExecutionException(
+                    String.format(
+                            "Could not find a result with result identifier '%s'.", resultId));
+        }
+        if (!result.isMaterialized()) {
+            throw new SqlExecutionException("Invalid result retrieval mode.");
+        }
+        return ((MaterializedResult) result).retrievePage(page);
+    }
+
+    public void cancelQuery(String resultId) throws SqlExecutionException {
+        DynamicResult result = resultStore.getResult(resultId);
+        if (result == null) {
+            throw new SqlExecutionException(
+                    String.format(
+                            "Could not find a result with result identifier '%s'.", resultId));
+        }
+
+        // stop retrieval and remove the result
+        LOG.info("Cancelling job {} and result retrieval.", resultId);
+        try {
+            // this operator will also stop flink job
+            result.close();
+        } catch (Throwable t) {
+            throw new SqlExecutionException("Could not cancel the query execution", t);
+        }
+        resultStore.removeResult(resultId);
+    }
+
+    // ---------------------------------------------------------------------------------------------
+    private <
+                    M extends MessageHeaders<R, P, U>,
+                    U extends MessageParameters,
+                    R extends RequestBody,
+                    P extends ResponseBody>
+            CompletableFuture<P> sendRequest(M messageHeaders, U messageParameters, R request)
+                    throws IOException {
+        return restClient.sendRequest(address, port, messageHeaders, messageParameters, request);
+    }
+
+    private FetchResultsResponseBody fetchResults(OperationHandle operationHandle) {
+        return fetchResults(operationHandle, 0L);
+    }
+
+    public FetchResultsResponseBody fetchResults(OperationHandle operationHandle, long token)
+            throws SqlClientException {
+        FetchResultsTokenParameters fetchResultsTokenParameters =
+                new FetchResultsTokenParameters(sessionHandle, operationHandle, token);
+        try {
+            return sendRequest(
+                            FetchResultsHeaders.getInstance(),
+                            fetchResultsTokenParameters,
+                            EmptyRequestBody.getInstance())
+                    .get();
+        } catch (Exception e) {
+            LOG.error(
+                    String.format(
+                            "Unexpected error occurs when fetching results. OperationHandle ID: %s.",
+                            operationHandle),
+                    e);
+            throw new SqlClientException(
```

Comment:
From a little more experimenting...  this line rethrows the exception and kills the SQL client process.

As a simple example, suppose one tries to `describe` a non-existent table.

With the local client, one sees:
```
Flink SQL> describe someTable;
[ERROR] Could not execute SQL statement. Reason:
org.apache.flink.table.api.ValidationException: Tables or views with the identifier 'default_catalog.default_database.someTable' doesn't exist```

In gateway mode, the result is a long exception: 

```
Flink SQL> describe someTable;                                                                                                                                                     
                                                                                                                                                                                   

        Exception in thread "main" org.apache.flink.table.client.SqlClientException: Unexpected error occurs when fetching execution results .                                             
        at org.apache.flink.table.client.gateway.remote.RemoteExecutor.executeStatement(RemoteExecutor.java:219)                                                                   
        at org.apache.flink.table.client.cli.CliClient.executeOperation(CliClient.java:667)                                                                                        
        at org.apache.flink.table.client.cli.CliClient.callOperation(CliClient.java:472)                                                                                           
        at org.apache.flink.table.client.cli.CliClient.executeOperation(CliClient.java:367)                                                                                        
        at org.apache.flink.table.client.cli.CliClient.getAndExecuteStatements(CliClient.java:323)                                                                                 
        at org.apache.flink.table.client.cli.CliClient.executeInteractive(CliClient.java:279)                                                                                      
        at org.apache.flink.table.client.cli.CliClient.executeInInteractiveMode(CliClient.java:227)                                                                                
        at org.apache.flink.table.client.SqlClient.openCli(SqlClient.java:148)                                                                                                     
        at org.apache.flink.table.client.SqlClient.start(SqlClient.java:96)                                                                                                        
        at org.apache.flink.table.client.SqlClient.startClient(SqlClient.java:170)                                                                                                 
        at org.apache.flink.table.client.SqlClient.main(SqlClient.java:158)                                                                                                        
Caused by: org.apache.flink.table.client.SqlClientException: Unexpected error occurs when fetching results. OperationHandle ID: 156c46b9-4278-4494-adb8-4339b8a71b20.
        at org.apache.flink.table.client.gateway.remote.RemoteExecutor.fetchResults(RemoteExecutor.java:389)                                      
        at org.apache.flink.table.client.gateway.remote.RemoteExecutor.fetchResults(RemoteExecutor.java:368)                                      
        at org.apache.flink.table.client.gateway.remote.RemoteExecutor.fetchWhenResultsReady(RemoteExecutor.java:406)                           
        at org.apache.flink.table.client.gateway.remote.RemoteExecutor.executeStatement(RemoteExecutor.java:212)               
        ... 10 more                                                                                                                                                                
Caused by: java.util.concurrent.ExecutionException: org.apache.flink.runtime.rest.util.RestClientException: [Internal server error., <Exception on server side:
org.apache.flink.table.gateway.api.utils.SqlGatewayException: org.apache.flink.table.gateway.api.utils.SqlGatewayException: Failed to fetchResults.
        at org.apache.flink.table.gateway.rest.handler.statement.FetchResultsHandler.handleRequest(FetchResultsHandler.java:86)                   
        at org.apache.flink.table.gateway.rest.handler.AbstractSqlGatewayRestHandler.respondToRequest(AbstractSqlGatewayRestHandler.java:84)      
        at org.apache.flink.table.gateway.rest.handler.AbstractSqlGatewayRestHandler.respondToRequest(AbstractSqlGatewayRestHandler.java:52)    
        at org.apache.flink.runtime.rest.handler.AbstractHandler.respondAsLeader(AbstractHandler.java:196)                            
        at org.apache.flink.runtime.rest.handler.LeaderRetrievalHandler.lambda$channelRead0$0(LeaderRetrievalHandler.java:83)                     
        at java.base/java.util.Optional.ifPresent(Optional.java:183)                                                                                                               
        at org.apache.flink.util.OptionalConsumer.ifPresent(OptionalConsumer.java:45)                                                                                              
        at org.apache.flink.runtime.rest.handler.LeaderRetrievalHandler.channelRead0(LeaderRetrievalHandler.java:80)
        at org.apache.flink.runtime.rest.handler.LeaderRetrievalHandler.channelRead0(LeaderRetrievalHandler.java:49)
        at org.apache.flink.shaded.netty4.io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99)
        at org.apache.flink.shaded.netty4.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at org.apache.flink.shaded.netty4.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at org.apache.flink.shaded.netty4.io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at org.apache.flink.runtime.rest.handler.router.RouterHandler.routed(RouterHandler.java:115)                                                                          
        at org.apache.flink.runtime.rest.handler.router.RouterHandler.channelRead0(RouterHandler.java:94)                           
        at org.apache.flink.runtime.rest.handler.router.RouterHandler.channelRead0(RouterHandler.java:55)                       
        at org.apache.flink.shaded.netty4.io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99)   
        at org.apache.flink.shaded.netty4.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at org.apache.flink.shaded.netty4.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at org.apache.flink.shaded.netty4.io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at org.apache.flink.shaded.netty4.io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)     
        at org.apache.flink.shaded.netty4.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at org.apache.flink.shaded.netty4.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at org.apache.flink.shaded.netty4.io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at org.apache.flink.runtime.rest.FileUploadHandler.channelRead0(FileUploadHandler.java:210)                                      
        at org.apache.flink.runtime.rest.FileUploadHandler.channelRead0(FileUploadHandler.java:69)                   
        at org.apache.flink.shaded.netty4.io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99)
        at org.apache.flink.shaded.netty4.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at org.apache.flink.shaded.netty4.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at org.apache.flink.shaded.netty4.io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at org.apache.flink.shaded.netty4.io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436)
        at org.apache.flink.shaded.netty4.io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:324)
        at org.apache.flink.shaded.netty4.io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:296)
        at org.apache.flink.shaded.netty4.io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251)
        at org.apache.flink.shaded.netty4.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at org.apache.flink.shaded.netty4.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at org.apache.flink.shaded.netty4.io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)    
        at org.apache.flink.shaded.netty4.io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
        at org.apache.flink.shaded.netty4.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at org.apache.flink.shaded.netty4.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at org.apache.flink.shaded.netty4.io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
        at org.apache.flink.shaded.netty4.io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
        at org.apache.flink.shaded.netty4.io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:719)
        at org.apache.flink.shaded.netty4.io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:655)
        at org.apache.flink.shaded.netty4.io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:581)
        at org.apache.flink.shaded.netty4.io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
        at org.apache.flink.shaded.netty4.io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)                   
        at org.apache.flink.shaded.netty4.io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at java.base/java.lang.Thread.run(Thread.java:829)                                                                                                                         
Caused by: org.apache.flink.table.gateway.api.utils.SqlGatewayException: Failed to fetchResults.                                 
        at org.apache.flink.table.gateway.service.SqlGatewayServiceImpl.fetchResults(SqlGatewayServiceImpl.java:203)            
        at org.apache.flink.table.gateway.rest.handler.statement.FetchResultsHandler.handleRequest(FetchResultsHandler.java:77)
        ... 48 more                                                                                                                                                                
Caused by: org.apache.flink.table.gateway.service.utils.SqlExecutionException: Failed to execute the operation 156c46b9-4278-4494-adb8-4339b8a71b20.
        at org.apache.flink.table.gateway.service.operation.OperationManager$Operation.processThrowable(OperationManager.java:391)
        at org.apache.flink.table.gateway.service.operation.OperationManager$Operation.lambda$run$0(OperationManager.java:249)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)   
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)                                                                                                      
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)          
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)                                                                                                      
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        ... 1 more
Caused by: org.apache.flink.table.api.ValidationException: Tables or views with the identifier 'default_catalog.default_database.someTable' doesn't exist
        at org.apache.flink.table.api.internal.TableEnvironmentImpl.executeInternal(TableEnvironmentImpl.java:1408)
        at org.apache.flink.table.gateway.service.operation.OperationExecutor.executeStatement(OperationExecutor.java:121)
        at org.apache.flink.table.gateway.service.SqlGatewayServiceImpl.lambda$executeStatement$0(SqlGatewayServiceImpl.java:182)
        at org.apache.flink.table.gateway.service.operation.OperationManager.lambda$submitOperation$1(OperationManager.java:111)
        at org.apache.flink.table.gateway.service.operation.OperationManager$Operation.lambda$run$0(OperationManager.java:240)
        ... 7 more

End of exception on server side>]
...
Shutting down the session...
done.
```

Reply:
Yes. I think we should introduce a new kind of `ErrorResponse` in the Gateway side. We can return the root cause, exception stack and error details.

Reply:
I am very sorry that when I propose this PR, my design of the Remote mode client has many defects. But I am sure those codes can be reused in my new design.

Reply:
Oh!  No need to be sorry; I just wanted to share what I have seen so far.

The embedded mode for the SQL client doesn't have the stack trace.  Maybe the client can default to showing the short error message (which may be corrective) and logging the stacktrace to a file?